src/Build/Evaluation/Evaluator.cs

```diff
@@ -1238,55 +1220,45 @@ private ICollection<P> AddToolsetProperties()
                 // set the property even if there is no matching sub-toolset.  
                 if (!_data.Properties.Contains(Constants.SubToolsetVersionPropertyName))
                 {
-                    P subToolsetVersionProperty = _data.SetProperty(Constants.SubToolsetVersionPropertyName, _data.SubToolsetVersion, false /* NOT global property */, false /* may NOT be a reserved name */);
-                    toolsetProperties.Add(subToolsetVersionProperty);
+                     _data.SetProperty(Constants.SubToolsetVersionPropertyName, _data.SubToolsetVersion, false /* NOT global property */, false /* may NOT be a reserved name */);
                 }
 
-                SubToolset subToolset;
-                if (_data.Toolset.SubToolsets.TryGetValue(_data.SubToolsetVersion, out subToolset))
+                if (_data.Toolset.SubToolsets.TryGetValue(_data.SubToolsetVersion, out SubToolset subToolset))
                 {
                     foreach (ProjectPropertyInstance subToolsetProperty in subToolset.Properties.Values)
                     {
-                        P property = _data.SetProperty(subToolsetProperty.Name, ((IProperty)subToolsetProperty).EvaluatedValueEscaped, false /* NOT global property */, false /* may NOT be a reserved name */);
-                        toolsetProperties.Add(property);
+                        _data.SetProperty(subToolsetProperty.Name, ((IProperty)subToolsetProperty).EvaluatedValueEscaped, false /* NOT global property */, false /* may NOT be a reserved name */);
                     }
                 }
             }
 
-            return toolsetProperties;
         }
 
         /// <summary>
         /// Put all the global properties into our property bag
         /// </summary>
-        private ICollection<P> AddGlobalProperties()
+        private int AddGlobalProperties()
         {
-            if (_data.GlobalPropertiesDictionary == null)
+            int count = 0;
+            if (_data.GlobalPropertiesDictionary != null)
             {
-                return Array.Empty<P>();
-            }
-
-            List<P> globalProperties = new List<P>(_data.GlobalPropertiesDictionary.Count);
-
-            foreach (ProjectPropertyInstance globalProperty in _data.GlobalPropertiesDictionary)
-            {
-                P property = _data.SetProperty(globalProperty.Name, ((IProperty)globalProperty).EvaluatedValueEscaped, true /* IS global property */, false /* may NOT be a reserved name */);
-                globalProperties.Add(property);
+                foreach (ProjectPropertyInstance globalProperty in _data.GlobalPropertiesDictionary)
+                {
+                    _data.SetProperty(globalProperty.Name, ((IProperty)globalProperty).EvaluatedValueEscaped, true /* IS global property */, false /* may NOT be a reserved name */);
+                    count++;
+                }
             }
-
-            return globalProperties;
+            return count;
```

Comment:
I hesitated briefly because it technically would change the behavior if there are already properties in that dictionary, which we could add, but then your way works and the current way doesn't, so it's a change we would make anyway. Thanks for the suggestion!